File: app/schemas/recommendation.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.parking_spot import SizeCategory, SpotType, VehicleType
from app.models.user import UserRole


class RecommendationWeights(BaseModel):
    availability: float = Field(default=0.35, ge=0, le=1)
    spot_type: float = Field(default=0.15, ge=0, le=1)
    zone: float = Field(default=0.10, ge=0, le=1)
    charger: float = Field(default=0.10, ge=0, le=1)
    role: float = Field(default=0.20, ge=0, le=1)
    conflict: float = Field(default=0.10, ge=0, le=1)


class RecommendationFilters(BaseModel):
    spot_types: list[SpotType] | None = None
    zone_ids: list[int] | None = None
    vehicle_type: VehicleType | None = None
    size_category: SizeCategory | None = None
    requires_charger: bool | None = None


class RecommendationPreferences(BaseModel):
    preferred_spot_types: list[SpotType] | None = None
    preferred_zone_ids: list[int] | None = None
    prefer_charger: bool = False
    needs_accessible_spot: bool = False
    max_results: int = Field(default=5, ge=1, le=50)


class RecommendationUserContext(BaseModel):
    role: UserRole | None = None


class RecommendationRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    user_context: RecommendationUserContext | None = None
    parking_lot_id: int
    from_time: datetime = Field(alias="from")
    to_time: datetime = Field(alias="to")
    filters: RecommendationFilters | None = None
    preferences: RecommendationPreferences | None = None
    weights: RecommendationWeights = Field(default_factory=RecommendationWeights)

    @model_validator(mode="after")
    def validate_interval(self):
        if self.from_time >= self.to_time:
            raise ValueError("'from' must be earlier than 'to'")
        return self


class RecommendationExplainFactor(BaseModel):
    factor: str
    value: float
    weight: float
    contribution: float
    reason: str


class DecisionFactor(BaseModel):
    name: str
    weight: float
    raw_value: float
    contribution: float
    explanation: str


class DecisionConstraint(BaseModel):
    name: str
    passed: bool
    explanation: str


class RejectedCandidate(BaseModel):
    spot_id: int
    reason: str
    constraint: str | None = None


class SelectedCandidate(BaseModel):
    spot_id: int
    spot_number: int
    spot_label: str
    final_score: float


class DecisionReport(BaseModel):
    selected_spot_id: int
    selected_spot_label: str
    final_score: float
    confidence: float
    factors: list[DecisionFactor]
    hard_constraints_passed: list[DecisionConstraint]
    rejected_candidates: list[RejectedCandidate]
    generated_at: datetime
    selected_candidate: SelectedCandidate


class RecommendedSpot(BaseModel):
    spot_id: int
    spot_number: int
    parking_lot_id: int
    zone_id: int | None
    zone_name: str | None
    spot_type: SpotType
    has_charger: bool
    score: float
    explainability: list[RecommendationExplainFactor]


class RecommendationResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    parking_lot_id: int
    from_time: datetime = Field(alias="from")
    to_time: datetime = Field(alias="to")
    requested_by_role: str
    total_candidates: int
    recommended_spots: list[RecommendedSpot]
    ranked_candidates: list[RecommendedSpot] = []
    rejected_candidates: list[RejectedCandidate] = []
    decision_report: DecisionReport | None = None

File: frontend/src/shared/types/recommendation.ts
import type { SizeCategory, SpotType, UserRole, VehicleType } from './common';

export interface RecommendationExplainFactor {
  factor: string;
  value: number;
  weight: number;
  contribution: number;
  reason: string;
}

export interface DecisionFactor {
  name: string;
  weight: number;
  raw_value: number;
  contribution: number;
  explanation: string;
}

export interface DecisionConstraint {
  name: string;
  passed: boolean;
  explanation: string;
}

export interface RejectedCandidate {
  spot_id: number;
  reason: string;
  constraint: string | null;
}

export interface SelectedCandidate {
  spot_id: number;
  spot_number: number;
  spot_label: string;
  final_score: number;
}

export interface DecisionReport {
  selected_spot_id: number;
  selected_spot_label: string;
  final_score: number;
  confidence: number;
  factors: DecisionFactor[];
  hard_constraints_passed: DecisionConstraint[];
  rejected_candidates: RejectedCandidate[];
  generated_at: string;
  selected_candidate: SelectedCandidate;
}

export interface RecommendedSpot {
  spot_id: number;
  spot_number: number;
  parking_lot_id: number;
  zone_id: number | null;
  zone_name: string | null;
  spot_type: SpotType;
  has_charger: boolean;
  score: number;
  explainability: RecommendationExplainFactor[];
}

export interface RecommendationResult {
  parking_lot_id: number;
  from: string;
  to: string;
  requested_by_role: UserRole | string;
  total_candidates: number;
  recommended_spots: RecommendedSpot[];
  ranked_candidates?: RecommendedSpot[];
  rejected_candidates?: RejectedCandidate[];
  decision_report?: DecisionReport | null;
}

export interface RecommendationRequestPayload {
  user_context?: {
    role?: UserRole;
  };
  parking_lot_id: number;
  from: string;
  to: string;
  filters?: {
    spot_types?: SpotType[];
    zone_ids?: number[];
    vehicle_type?: VehicleType;
    size_category?: SizeCategory;
    requires_charger?: boolean;
  };
  preferences?: {
    preferred_spot_types?: SpotType[];
    preferred_zone_ids?: number[];
    prefer_charger?: boolean;
    needs_accessible_spot?: boolean;
    max_results?: number;
  };
  weights?: {
    availability?: number;
    spot_type?: number;
    zone?: number;
    charger?: number;
    role?: number;
    conflict?: number;
  };
}
